REST_API_COMPATIBILITY.md

```diff
@@ -0,0 +1,182 @@
+#  REST API compatibility developers guide
+
+REST API compatibility is intended to minimize the impact for breaking changes applied to the REST API. Compatibility is implemented is a best effort to strike a balance between preserving the REST API contract across major versions while still allowing for breaking changes. URL paths, parameters, HTTP verbs, request bodies and response bodies are all covered by REST API compatibility.
+
+### Example use case
+
+The recommended procedure to upgrade Elasticsearch to a new major version is to first upgrade Elasticsearch (the server), then upgrade any of the consumers/clients. This implies that the consumers/clients during and immediately after an upgrade may still be attempting to communicate with Elasticsearch using the prior major version's REST API contract.  During this time if any of consumers/clients continue to call into one of the REST API's that include breaking changes, the consumer/client can also break.  Compatibility is a best attempt to intended honor the spirit of the original API call even after the breaking change has been applied.
```

Comment:
I am not sure I understand.
```
During this time if any of consumers/clients continue to call into one of the REST API's that include breaking changes, the consumer/client can also break.
```
if the client is using previous version with Compatible API, then it should not break (in most cases)